unixfs/README.md

```diff
@@ -37,6 +37,16 @@ message Data {
 
 	optional uint64 hashType = 5;
 	optional uint64 fanout = 6;
+	
+	/* file metadata */
+	optional uint32 mtime = 7;
+	optional uint32 atime = 8;
+	optional uint32 ctime = 9;
+	
+	/* file permissions */
+	optional uint32 permissions = 10;
```

Comment:
Instead of `permissions`, I recommend calling this`filemode` and have it be the literal spec/32-bit representation of golang's `FileMode`. It already [captures the interesting "data bits" really well](https://golang.org/pkg/os/#FileMode) for our use cases, and is also a stable format: https://github.com/golang/go/issues/25422#issuecomment-524078451

Reply:
Things that aren't files have permissions too, so I don't think `filemode` is a great name.

`permissions` is the name everyone uses when talking about permissions so it seems fine to me.

If not that then how about `accessmode`?

Reply:
> permissions is the name everyone uses when talking about permissions

I am not only talking about permissions though :) I am talking about the entirety of `mode_t` encoded as a 32bit field. The nomenclature is a bit confusing, but this is POSIX for you: among other things a **`directory`** is one of the [available **file** types](https://www.gnu.org/software/libc/manual/html_node/Testing-File-Type.html#Testing-File-Type).

The reason it is extremely interesting to have the filetype info on each individual entry ( just like in common VFS `inode`s ) is so that after fetching a directory block you can **instantly** tell how any files, subdirectories, links and other weirdness it contains. With no further trips to the network.

Reply:
How about just `mode`?

Reply:
I like `mode` as a name.  :+1: 

I'd refrain from referring directly to the golang filemode bits for docs.  Three reasons: They've combined several other things in that struct which are not sensible for us to combine (namely, file _types_, which we represent with different messages entirely, right?).  They've used a very unusual bit position for the 'sticky' bit (this has caused me painful bugs in the past).  And finally, it's probably easier to just write our own doc labeling each bit than it is to outsource.

We also need to document what is and is not in this field.  Either "permissions" or the "mode" name are very likely to carry the connotations of more than just the 0777/rwxrwxrwx bits for many readers.

The typical bits beyond the 0777 mask are: setuid, setguid, and sticky.  We should document clearly if we're doing those or not, and what position they're in. In other systems, these (again, 'typically') pack into the 07000 mask.  [I've done digging on this before](https://github.com/polydawn/rio/blob/4fef789ffc851aaf3ac291a91c21e477e6367811/fs/metadata.go#L29-L50) and if we're going to do these as a bitmask, I'd hew to the common form used there.  But we shouldn't skip the question of if we're doing them at all; and the question of if we're doing them as a bitmask.

My personal weakly-held opinion is to advocate for considering support for these additional bits, but not including them in a bitmask.  None of these three additional bits is frequently used in bulk (whole computers generally have a few dozen setuid/setgid things on the whole filesystem, and approximately *one* sticky bit (on `/tmp`)), so there's no reason to work about minimizing their encoded size; and they're also *extremely* important bits to Handle Carefully because of their relationship to system security, so setting them apart and handling them with oven mitts so to speak is a good idea.

Reply:
changed to `mode`.

Reply:
I’ve added a note about `mode` being in standard numeric notation and linked to the wikipedia article describing the behavior. https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation

Reply:
The official POSIX definition is: https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/sys_stat.h.html.

I'd say that only the lower 9 bits (ugo, rwx) are defined for now.

Reply:
I’ll add this to the comment as well.